### PR TITLE
compile integration tests as a single binary

### PR DESCRIPTION
**Pull request by jyn514**
closes https://github.com/martinvonz/jj/issues/2912

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/tests/runner.rs
Comment (yuja) [inline lib/tests/runner.rs:2]: nit: this looks a bit cryptic.

Perhaps, the inner function can be extracted to `testutils`? Then we can add a thin `#[test]` that calls the function with the runner file path?

**Review by yuja — COMMENTED** [review]
While this is a bit adhoc, I like the end result.

@arxanas any thoughts?

File: cli/tests/test_no_abandoned_test_files.rs
Comment (ilyagr) [inline cli/tests/test_no_abandoned_test_files.rs:5]: Optional: If that's easy, I'd consider putting this directly into `runner.rs`.

**Update:** Yuya's suggestion in https://github.com/martinvonz/jj/pull/2914/commits/5daccacfd890cae1c6161ce0a76c9da0c90edca9#r1476082844 sounds good to me as well.

File: cli/tests/test_no_abandoned_test_files.rs
Comment (ilyagr) [inline cli/tests/test_no_abandoned_test_files.rs:17]: Nit: I'd search directly for `mod {name};` here, to match the error for example.

**Review by ilyagr — COMMENTED** [review]
Wow, this is fast 🚀. I might no longer have to be paranoid about linking speed and all the ways phases of the moon that can affect it. LGTM, but I'll let others officially approve.

Aside: I wish Rust more obviously encouraged this style, or defaulted to building one binary for the tests.

File: cli/tests/test_no_abandoned_test_files.rs
Comment (ilyagr) [reply cli/tests/test_no_abandoned_test_files.rs:5]: If it's not easy, I'd put a comment next to this test in `runner.rs`. If you don't, I can do it myself after this is merged.

**Comment by arxanas** [comment]
Can you compare the resulting runtimes for running the whole test suite and running a single test? The improved compilation times are admirable, but if it turns out that the full test suite :: 5m -> 4m but incremental tests :: 2s -> 4s then it might not be worth it. Ideally I would have incremental rebuild+test times be <1s. When developing jj, I extensively relied on re-running individual tests repeatedly, so that feedback loop is very important to me. Others might feel differently about those trade-offs.

**Comment by ilyagr** [comment]
The benchmark in the commit message is `touch lib/src/lib.rs ; cargo t --test runner --no-run`, and I'm assuming `hyperfine` throws out the first run like a good benchmarking tool should. So, it's all about incremental compilation, I think.

**Comment by ilyagr** [comment]
I also unscientifically tried `touch lib/src/lib.rs ; cargo insta test --test-runner nextest -- test_abandon`. It feels fast with this PR, though which linker you use might again make a difference.

I found another minor downside: the fish shell has completions for `cargo test --test <TAB>`, but not for `cargo test -- <TAB>` (or they don't work with this PR). In this way, doing a subset of tests is less convenient for me with this PR (though I think the speedup is worth it for me).

**Comment by arxanas** [comment]
Regarding `touch`, @jyn514 would probably know better than me, but in my blog post I actually modified the file in the worry that `cargo` might do something content-addressed as part of compilation and it wouldn't represent a "real" incremental change.

**Comment by jyn514** [comment]
> Regarding `touch`, @jyn514 would probably know better than me, but in my blog post I actually modified the file in the worry that `cargo` might do something content-addressed as part of compilation and it wouldn't represent a "real" incremental change.

cargo does not do content-addressed hashing, no. you can see this if you set CARGO_LOG=info:
```
; touch lib/src/lib.rs ; CARGO_LOG=info cargo t --no-run
   0.089419238s  INFO cargo::core::compiler::fingerprint: stale: changed "/home/jyn/src/jj/lib/src/lib.rs"
   0.089482460s  INFO cargo::core::compiler::fingerprint:           (vs) "/home/jyn/.local/lib/cargo/target/debug/.fingerprint/jj-lib-49a46e34c360fc2e/dep-lib-jj-lib"
   0.089485415s  INFO cargo::core::compiler::fingerprint:                FileTime { seconds: 1706861032, nanos: 988871571 } < FileTime { seconds: 1706931264, nanos: 137617386 }
```

> Can you compare the resulting runtimes for running the whole test suite and running a single test? The improved compilation times are admirable, but if it turns out that the full test suite :: 5m -> 4m but incremental tests :: 2s -> 4s then it might not be worth it. Ideally I would have incremental rebuild+test times be <1s. When developing jj, I extensively relied on re-running individual tests repeatedly, so that feedback loop is very important to me. Others might feel differently about those trade-offs.

sure; but the times are small enough i don't think times on my computer will be comparable to yours, i'd recommend running this on your own machine to see what the times are.

before this change:
```
; hyperfine 'touch lib/src/lib.rs && cargo t --test test_working_copy'
Benchmark 1: touch lib/src/lib.rs && cargo t --test test_working_copy
  Time (mean ± σ):      3.543 s ±  0.168 s    [User: 2.597 s, System: 1.262 s]
  Range (min … max):    3.400 s …  3.847 s    10 runs
```

after this change:
```
; hyperfine 'touch lib/src/lib.rs && cargo t --test runner -- test_working_copy'
Benchmark 1: touch lib/src/lib.rs && cargo t --test runner -- test_working_copy
  Time (mean ± σ):      4.129 s ±  0.120 s    [User: 3.636 s, System: 1.593 s]
  Range (min … max):    3.933 s …  4.346 s    10 runs
```

so only about half a second of regression, from 3.5 to 4 seconds.

(using `cargo test` is intentional, nextest is actually a bit *slower* than the built-in framework for single tests)

**Comment by jyn514** [comment]
> I found another minor downside: the fish shell has completions for `cargo test --test <TAB>`, but not for `cargo test -- <TAB>` (or they don't work with this PR). In this way, doing a subset of tests is less convenient for me with this PR (though I think the speedup is worth it for me).

yeah, this is unfortunate. i don't know a way to fix this upstream in fish; it would have to parse the `mod` declarations (which is not *super* hard but requires it to be aware of cargo's directory structure instead of just running `cargo metadata`).

**Comment by jyn514** [comment]
> Can you compare the resulting runtimes for running the whole test suite

it takes about 13.5 seconds for me to run the tests after they're compiled, both before and after this PR. so the total time is going from ~45 to ~18 seconds (and it's better than 18 in practice because it's rare for the very last test run to be the only one that fails).

File: lib/tests/runner.rs
Comment (jyn514) [reply lib/tests/runner.rs]: great idea, done! that also lets me move the `#[test]` to `runner.rs` as @ilyagr suggested :)

File: cli/tests/runner.rs
Comment (ilyagr) [inline cli/tests/runner.rs:6]: Non-blocking: Perhaps `no_forgotten_test_files` (here and in the testutils function) is a little better?

I'm not completely happy with that name either, but we can always change the name later.

File: cli/tests/test_global_opts.rs
Comment (ilyagr) [inline cli/tests/test_global_opts.rs:19]: This might be a false alarm, but it seems to me that this PR causes the `jj-cli::runner test_global_opts::test_no_subcommand` test to use the system `jj` instead of `target/debug/jj` half the time. This seems non-deterministic.

I'm running `cargo insta test --test-runner nextest --workspace --accept`, which is essentially `cargo nextest run`. I'm guessing that whether the problem occurs depends on the order of test execution.

The reason I noticed is that my system jj has https://github.com/martinvonz/jj/pull/2034 baked-in, which causes this test to fail unless it's appropriately modified. I guess a better way to debug would be to create `/tmp/bug/jj` that always panics and run `PATH=/tmp/bug:$PATH cargo nextest run`, which I'll try to do.

Comment (ilyagr) [reply cli/tests/test_global_opts.rs:19]: Never mind, that has nothing to do with your PR. The problem occurs when running tests in a merge commit.

Comment (jyn514) [reply cli/tests/test_global_opts.rs:19]: i haven't been able to replicate that failure. i do get occasional flakes from `test_gc`, but i would be surprised if that's related to this change?
```
--- STDERR:              jj-lib::runner test_git_backend::test_gc ---
thread 'test_git_backend::test_gc' panicked at lib/tests/test_git_backend.rs:151:5:
assertion `left == right` failed
  left: {CommitId("577f8238c5611e318786cad886bfe6b479538687"), CommitId("7d52d6a9cebe821d379c2d1d5a707f04a3dd25f2")}
 right: {CommitId("da30a5e5f1c460afabd6f2594a2a7da8df039d73"), CommitId("577f8238c5611e318786cad886bfe6b479538687"), CommitId("7d52d6a9cebe821d379c2d1d5a707f04a3dd25f2")}
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
looking at https://docs.rs/assert_cmd/latest/src/assert_cmd/cargo.rs.html#183-208, it's deterministic; even if the binary doesn't exist yet, it returns an error rather than falling back to PATH. so i'm not sure how this could be related.

Comment (ilyagr) [reply cli/tests/test_global_opts.rs:19]: The problem also has nothing to do with the system `jj`. Sometimes (non-deterministically), jj's version includes a concatenation of the merge commit's parents' commit ids. I'm guessing that happens when `build.rs` uses git instead of jj to get the version. I have no idea why it's non-deterministic.

Comment (ilyagr) [reply cli/tests/test_global_opts.rs:19]: I created a separate bug for this: https://github.com/martinvonz/jj/issues/2922

**Comment by martinvonz** [comment]
Thanks for the PR, @jyn514, and thanks for reviewing and testing to everyone else. @arxanas, are you okay with this? It sounds like we're just waiting for your approval, but no rush.

**Review by martinvonz — APPROVED** [review]
I'm giving my approval now anyway. It sounds like most people are for this change. We can always roll it back if it turns out we missed something. Thanks again, jyn!

**Comment by ilyagr** [comment]
I think Martin will cut a release in the next couple of days. I'm likely being overly cautious, but perhaps we can delay merging this until the release, just in case?

**Comment by martinvonz** [comment]
> I think Martin will cut a release in the next couple of days. I'm likely being overly cautious, but perhaps we can delay merging this until the release, just in case?

It doesn't seem that this can break the binaries, so what's your concern? If it breaks the CI actions, I'd rather learn that tomorrow than in a month.

**Comment by ilyagr** [comment]
OK, this is definitely up to Martin. Please feel free to merge, and thank you again!

**Pull request merged**
